internal/controller/phases.go

```diff
@@ -279,6 +282,41 @@ func (p *phaseReconciler) configmapRepository(ctx context.Context, labelSelector
 	return mr, nil
 }
 
+// getComponentsData returns components data based on if it's compressed or not.
+func getComponentsData(cm corev1.ConfigMap) (string, error) {
+	// Data is not compressed, return it immediately.
+	if cm.GetAnnotations()[compressedAnnotation] != "true" {
+		components, ok := cm.Data[componentsConfigMapKey]
+		if !ok {
+			return "", fmt.Errorf("ConfigMap %s/%s Data has no components", cm.Namespace, cm.Name)
+		}
+
+		return components, nil
+	}
+
+	// Otherwise we have to decompress the data first.
```

Comment:
Q to better understand this logic: what is the use case when it is already compressed and why we need to decompress in that case?

Reply:
In the normal workflow, when we create a provider for the first time, we download manifests from the provided URL and store them in a configmap to reuse later. Unfortunately, there is a size limit (1MiB) on the configmap in kubernetes, and for some providers we can't store there manifests directly.

Here is an example of capi provider that has infrastructure-components.yaml file exceeding 1MiB: https://github.com/oracle/cluster-api-provider-oci/releases/tag/v0.12.0

The error looks like this:
```
E0718 10:19:28.002488       1 controller.go:329]  "msg"="Reconciler error" "error"="failed to create config map for provider \"oci\": ConfigMap \"infrastructure-oci-v0.12.0\" is invalid: []: Too long: must have at most 1048576 bytes" "InfrastructureProvider"={"name":"oci","namespace":"cluster-api"} "controller"="infrastructureprovider" "controllerGroup"="operator.cluster.x-k8s.io" "controllerKind"="InfrastructureProvider" "name"="oci" "namespace"="cluster-api" "reconcileID"="
3530ba70-285c-4a14-997b-5bdfc092c973
```

To workaround this we put compressed data in the configmap, which is 10 times smaller than the original. Later, when we generate clusterctl provider repository in the "load" phase, we decompress it and use as before.

Reply:
Thanks @Fedosin!